Update 17.0.6

danpetry opened the pull request:
**Destination channel:** defaults

### Links

- [PKG-3629](https://anaconda.atlassian.net/browse/PKG-3629) 
- [Upstream repository](https://github.com/llvm/llvm-project/tree/llvmorg-17.0.6/mlir)
- [Conda-forge recipe](https://github.com/conda-forge/mlir-feedstock/blob/main/recipe/meta.yaml)

### Explanation of changes:

- build for python 3.12
- include some changes from conda-forge's recipe
- Add tests at the end of build.sh (at the moment they fail, see jira ticket above for further work on this, possibly to be done in a future ticket)
- clean out some unused stuff: patches, environment variables, deps
- various fixes to get it building
- changes to address lint checks

[PKG-3629]: https://anaconda.atlassian.net/browse/PKG-3629?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

danpetry committed:
Add changes from conda-forge

danpetry committed:
update to 17.0.4

danpetry committed:
Clean out the conda-forge stuff

danpetry committed:
fix indentation

danpetry committed:
Various tidying-up changes

- Quoting/escaping test strings
- Removing unused patches
- removing some other stuff we don't need

danpetry committed:
Use llvmdev most recent staging channel

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:22: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:40: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:75: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:92: missing_tests: output "libmlir17": No tests were found.
ERROR: clone/recipe/meta.yaml:105: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:108: should_use_compilers: The recipe requires a compiler directly
WARNING: clone/recipe/meta.yaml:123: missing_description: The recipe is missing a description
WARNING: clone/recipe/meta.yaml:126: missing_description: The recipe is missing a description
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:22: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:40: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:75: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:92: missing_tests: output "libmlir17": No tests were found.
ERROR: clone/recipe/meta.yaml:105: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:108: should_use_compilers: The recipe requires a compiler directly
WARNING: clone/recipe/meta.yaml:123: missing_description: The recipe is missing a description
WARNING: clone/recipe/meta.yaml:126: missing_description: The recipe is missing a description
Errors were found

</details>

danpetry committed:
A few bugfixes to address conda-build errors

Specifically, missing run dep on mlir and missing ld64.so.1 on s390x

Also removed redundant jinja

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:22: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:40: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:78: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:95: missing_tests: output "libmlir17": No tests were found.
ERROR: clone/recipe/meta.yaml:111: should_use_compilers: The recipe requires a compiler directly
WARNING: clone/recipe/meta.yaml:129: missing_description: The recipe is missing a description
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:22: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:40: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:78: should_use_compilers: The recipe requires a compiler directly
ERROR: clone/recipe/meta.yaml:95: missing_tests: output "libmlir17": No tests were found.
ERROR: clone/recipe/meta.yaml:111: should_use_compilers: The recipe requires a compiler directly
WARNING: clone/recipe/meta.yaml:129: missing_description: The recipe is missing a description
Errors were found

</details>

danpetry committed:
Remove staging channel because llvm is released now

danpetry committed:
Install libmlir_float16_utils

danpetry committed:
Add missing libc++ dependency

danpetry committed:
Some tentative changes to try and address current errors

danpetry committed:
On unix, copy in libraries to prefix rather than deleting all libs in prefix

danpetry committed:
delete trailing whitespace

danpetry committed:
a couple of bugfixes

danpetry committed:
Some more bugfixes

danpetry committed:
More bugfixes

danpetry committed:
Adjust unix compiler options

danpetry committed:
Update build flags

danpetry committed:
enable testing

danpetry committed:
Adjust c++ stuff, debugging lit testing

danpetry committed:
lint'er

danpetry committed:
Adding tests, debugging test runner

danpetry committed:
A bunch of scrappy attempts to get things working

danpetry committed:
Revert "A bunch of scrappy attempts to get things working"

This reverts commit 2bae9734fcc3af90c78ea5f6597b9cfb7e325e2b.

danpetry committed:
Add patch to find correct .dylibs, on mac at least

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_e8__zbl91n/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:135: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
update patch so it doesn't reference a variable that doesn't exist

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_0a8hfr95_2/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:135: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
Get tests running successfully

- use the correct flag to CMake configure step to find the .dylibs
- copy required (rendered) config file from build to test directory

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_49b6wyzhii/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:132: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
Revert "Clean out the conda-forge stuff"

This reverts commit a7e149d6b098810d7e0696f24f20caa499847d95.

danpetry committed:
Don't stop failing tests from passing build. to be fixed later

danpetry committed:
Add flag to reflect triton's llvm build flag

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_733dqwdnwi/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:132: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
Previous skipping test method didn't work, try something else

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_63mqd7qwgd/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:132: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
try again

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_f36nug_b78/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:132: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
Comment out broken win-64 test (to be fixed later)

danpetry committed:
Remove unused CMake variables

danpetry committed:
Remove zlib; overdepending checks say it's not being used

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_a88wihqfv0/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:128: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

danpetry committed:
Remove patch series file added by accident

danpetry committed:
Provide zlib on windows, it's required

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_7f3294is77/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:127: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
- clone/recipe/meta.yaml:128: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

JeanChristopheMorinPerso commented on recipe/0001-Support-cross-compiling-standalone-MLIR.patch:
Removing files fro conda-forge makes it harder to sync the feedstock later. In this case, the patch seems harmless and we could probably build with it without any undesirable side-effects (I think).

danpetry replied:
fair enough, yeah. Although my two cents, our recipes tend to deviate significantly from conda-forge, so it's not feasible to sync them, very often. And anyway, we want to understand the changes they made rather than just pull them across

JeanChristopheMorinPerso replied:
>our recipes tend to deviate significantly from conda-forge, so it's not feasible to sync them, very often.

Yeah, indeed. But we shouldn't make it even harder to sync them IMO.

>And anyway, we want to understand the changes they made rather than just pull them across

I 100% agree. Understanding what their patches are doing is important.

JeanChristopheMorinPerso commented on recipe/bld.bat:
```diff
@@ -1,14 +1,30 @@
 if %ERRORLEVEL% neq 0 exit 1
+
+REM this is where lit expects to find helper tools. Perhaps they should be put here while building llvm.
+cp %PREFIX%\libexec\llvm\* %PREFIX%\bin
```
Should we remove these after the tests? If we don't they'll be ship with the package...

danpetry replied:
done

danpetry replied:
actually there wasn't anything being copied. Maybe why the tests weren't running.

JeanChristopheMorinPerso commented on recipe/build.sh:
```diff
@@ -15,11 +20,34 @@ cd build
+  -DMLIR_INCLUDE_TESTS=ON \
   -DLLVM_BUILD_LLVM_DYLIB=ON \
+  -DMLIR_INCLUDE_INTEGRATION_TESTS=ON \
+  -DLLVM_SHLIB_OUTPUT_INTDIR="${SRC_DIR}/build/lib" \
```
Should we add a comment here considering we spent a couple of hours to figure this one out?

danpetry replied:
done

cbouss commented on recipe/meta.yaml:
```diff
@@ -75,44 +78,59 @@ outputs:
-        - {{ pin_subpackage("libmlir" + major_version, exact=True) }}  # [unix]
+        - {{ pin_subpackage("libmlir" ~ major_version, exact=True) }}  # [unix]
+        - libcxx >={{ cxx_compiler_version }}  # [osx]
+        - zlib # [win]
```
zlib and zstd should be included in the host section, otherwise the run_exports won't be applied.
They should also be included in the global host section, as the build happens at the global level.

danpetry replied:
I'm not following - isn't putting it in the run section effectively the same? It's not needed during build time, it just fails the linking check if this isn't here

JeanChristopheMorinPerso replied:
If it fails at runtime, doesn't it mean that it's being linked at build time?

danpetry replied:
yep, charles just explained that it is probably being provided by some other dependency. I thought the overlinking check would get that case, but I guess overlinking relates to the run environment, not the host one?

JeanChristopheMorinPerso replied:
>I guess overlinking relates to the run environment, not the host one

I think so yes.

danpetry replied:
As a related question, if a package is included in the host section, does it pull in its dependencies from its recipe's _run_ section into the host environment? (I think so - same for if its included in the build section too?)

JeanChristopheMorinPerso replied:
Yep. If you include a package in any section, the transitive dependencies will also be installed. If this wasn't the case, package managers wouldn't be very useful 😛

JeanChristopheMorinPerso commented on recipe/build.sh:
```diff
@@ -15,11 +20,34 @@ cd build
 
-ninja
+cmake --build . -- -j${CPU_COUNT}
+#echo ${BUILD_PREFIX}
+#ls -la "${BUILD_PREFIX}/bin"
+# is PREFIX the env or host directory?
+#echo ${PREFIX}
```
```suggestion
```
Should this be removed?

JeanChristopheMorinPerso commented on recipe/build.sh:
```diff
@@ -15,11 +20,34 @@ cd build
+#echo ${PREFIX}
+
+# this is where lit expects to find helper tools. Perhaps they should be put here while building llvm.
+cp ${PREFIX}/libexec/llvm/* ${PREFIX}/bin
```
Same here, should we remove files copied in `${PREFIX}/bin` after the tests are run?

danpetry replied:
done

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -75,44 +78,59 @@ outputs:
       host:
         - llvmdev {{ version }}
         - llvm {{ version }}
-        - zlib
```
We don't need zlib?

danpetry replied:
apparently not. Build succeeds without. I'm not sure why conda-forge has it in

danpetry committed:
Restore patch from conda-build for closer parity with them

danpetry committed:
Add comment explaining the cmake arg required to run the tests

danpetry committed:
Remove unused commented code

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_18iyc60lh5/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:127: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
- clone/recipe/meta.yaml:128: cbc_dep_in_run_missing_from_host: Run dependencies listed in the cbc should also be present in the host section.
===== Final Report: =====
2 Errors and 0 Warnings were found

</details>

danpetry committed:
Clean up tools required by lit so they don't get packaged

danpetry committed:
Move zlib and zstd into host, so the build uses the ones supplied by this recipe

danpetry committed:
Rebase patch

danpetry committed:
Another go at rebasing

danpetry committed:
Remove nonexistent tool from end of list

JeanChristopheMorinPerso reviewed (approved):
This lookg good to me. Awesome job on this one @danpetry !

danpetry commented:
Thanks 😺

skupr-anaconda commented on recipe/meta.yaml:
```diff
@@ -24,12 +22,18 @@ requirements:
+    # both of these required to run testing
+    - lit     {{ version }}
+    - llvmdev {{ version }}
+    - python
```
`ninja` already contains `python`, see https://github.com/AnacondaRecipes/ninja-feedstock/blob/02e0641538afcc8657eeeea966b1c0c672f6231d/recipe/meta.yaml#L36

danpetry replied:
I put this as a to-do for next time if that's ok. (Is python expressed as a dep explicitly upstream?)

skupr-anaconda reviewed (approved):
LGTM

cbouss reviewed (approved):
LGTM, nice work!
